src/server/pps/persist/server/rethink_api_server.go

```diff
@@ -320,6 +319,41 @@ func (a *rethinkAPIServer) DeletePipelineInfo(ctx context.Context, request *ppsc
 	return google_protobuf.EmptyInstance, nil
 }
 
+func (a *rethinkAPIServer) ShardStart(ctx context.Context, request *ppsclient.Job) (response *persist.JobInfo, retErr error) {
+	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
+	return a.shardOp(ctx, request, "ShardsStarted")
+}
+
+func (a *rethinkAPIServer) ShardSucceed(ctx context.Context, request *ppsclient.Job) (response *persist.JobInfo, retErr error) {
+	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
+	return a.shardOp(ctx, request, "ShardsSucceeded")
+}
+
+func (a *rethinkAPIServer) ShardFail(ctx context.Context, request *ppsclient.Job) (response *persist.JobInfo, retErr error) {
+	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
+	return a.shardOp(ctx, request, "ShardsFailed")
+}
+
+func (a *rethinkAPIServer) shardOp(ctx context.Context, request *ppsclient.Job, field string) (response *persist.JobInfo, retErr error) {
+	cursor, err := a.getTerm(jobInfosTable).Get(request.ID).Update(map[string]interface{}{
+		field: gorethink.Row.Field(field).Add(1).Default(0),
+	}, gorethink.UpdateOpts{
+		ReturnChanges: true,
+	}).Field("changes").Field("new_val").Run(a.session)
+
```

Comment:
This is kind of pedantic but I think it's good form to not put a newline between error declartion and checking since they're really part of the same operation. Would you be cool with that?